pages/add/add.jsx

```diff
@@ -1,12 +1,201 @@
+  },
+  componentDidMount() {
+    let loggedInParam = this.props.router.location.query.loggedin;
+
+    if ( !loggedInParam || loggedInParam.toString().toLowerCase() === `true` ) {
+      // if there's no "loggedin" param presented in the URL OR if "loggedin" param is true
+      // we want to verify if user is really logged in by making a /userstatus call to Pulse API
+      this.verifyLoggedInStatus();
+    } else if ( loggedInParam && loggedInParam.toString().toLowerCase() === `false` ) {
+      // if "loggedin" param is presented in the URL and it is set to false
+      // we assume that user cannot be logged in using his/her Google Account
+      // (i.e., not using a Mozilla Staff account)
+      this.updateLoggedInInfo(false);
+    }
+  },
+  verifyLoggedInStatus() {
+    Service.userstatus()
+      .then(response => {
+        if (response.loggedin) {
+          // Pulse API verified that this user was already logged in
+          this.updateLoggedInInfo(true,response.username);
+        } else {
+          // user is not logged in, redirect him/her to Pulse API's login URL
+          let redirectUrl = `${env.ORIGIN}${this.props.router.getCurrentLocation().pathname}`;
+
+          window.location.href = UserData.logInUrl.get(redirectUrl);
+        }
+      })
+      .catch(reason => {
+        this.updateLoggedInInfo(false);
+        console.error(reason);
+      });
+  },
+  updateLoggedInInfo(status, username = ``) {
+    UserData.userLoggedInStatus.set(status);
+    UserData.username.set(username);
+
+    // we have verified user's logged in status with Pulse API
+    this.setState({
+      loggedInStatusChecked: true
+    });
+
+    // After the login process user will be redirected back to Pulse.
+    // There will be a query argument "loggedin" presented in the URL
+    // which is something we don't want it to stay at all times.
+    this.clearLoggedinParam();
+  },
+  clearLoggedinParam() {
+    let query = this.props.router.location.query;
+
+    // we don't need the "loggedin" param sent back from Pulse API to be presented at all times
+    delete query.loggedin;
+
+    browserHistory.replace({
+      pathname: this.props.router.location.pathname,
+      query: query
+    });
+  },
+  logOutUser(event) {
+    event.preventDefault();
+    Service.logout()
+      .then(() => {
+        // we have successfully logged user out
+        this.updateLoggedInInfo(false);
+        // redirect user to /featured page
+        browserHistory.push({
+          pathname: `/featured`
+        });
+      })
+      .catch(reason => {
+        console.error(reason);
+        // TODO:FIXME: what do we do here?
+      });
+  },
+  handleCreatorClick(event) {
+    event.preventDefault();
+    this.setState({numCreatorFields: this.state.numCreatorFields+1});
+  },
   render() {
+    let creatorFields = [];
+
+    for (var i=0; i<this.state.numCreatorFields; i++) {
+      creatorFields.push( <CreatorInputField key={i} />);
```

Comment:
This is actually a symptom of an architectural flaw: either this is a pure HTML form, in which case this component should be stateless, or this is a data UI, in which case the component should track all the form data in its state, and the form should have `onChange` handling for each field so that things like dynamically adding/removing creator fields with proper keys is actually possibe.

Right now you've written the form in a way that hands off all the functionality to the browser instead of React, and so if the component needs to rerender for whatever reason, the form as the user has it filled in so far will get emptied. So, if we want to be able to add more creators for something with more than three creators, this would need to be rewritten.

So the main question is: how much work do we want to do now, vs. later? We can present a static form with three creator fields, and then we should hard code that, and remove any kind of state from this component: it doesn't have one. 

Or, if we want a variable number of creators, then we need to rewrite this to be a proper React form based off of state for all the fields we care about, in which case we want to use a form builder for that (we have one over on https://github.com/mozilla/learning.mozilla.org/blob/master/pages/clubs/form/builder/Form.jsx but there might be others on npm these days, too).